daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -816,32 +827,33 @@ private[lf] object Speedy {
       try {
         @tailrec
         def loop(): SResult[Q] = {
-          if (enableInstrumentation) {
+          if (enableInstrumentation)
             Classify.classifyMachine(this, track.classifyCounts)
-          }
-          if (enableLightweightStepTracing) {
-            steps += 1
-            println(s"$steps: ${PrettyLightweight.ppMachine(this)}")
-          }
           val thisControl = control
           setControl(Control.WeAreUnset)
-          thisControl match {
-            case Control.WeAreUnset =>
-              sys.error("**attempt to run a machine with unset control")
-            case Control.Expression(exp) =>
-              control = exp.execute(this)
-              loop()
-            case Control.Value(value) =>
-              popTempStackToBase()
-              control = popKont().execute(this, value)
-              loop()
-            case Control.Question(res) =>
-              SResultQuestion(res)
-            case Control.Complete(value: SValue) =>
-              if (enableInstrumentation) track.print()
-              SResultFinal(value)
-            case Control.Error(ie) =>
-              SResultError(SErrorDamlException(ie))
+          if (interruptionCountDown == 0) {
+            interruptionCountDown = iterationsBetweenInterruptions
+            SResultInterruption(() => control = thisControl)
```

Comment:
Remove the callback parameter from SResultInterruption.
Move the assignment to `control` to just after the assignment to `interruptionCountdown`.